systemd: 245.6 -> 246

flokli opened the pull request:
###### Motivation for this change
This bumps systemd to its latest version, containing many fixes. See https://github.com/systemd/systemd/blob/v246/NEWS for release notes.

This is still WIP - `libsystemd.so.0.29.0` from the `lib` output refers back to the `out` output, causing a cyclic dependency.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

flokli committed:
systemd: 245.7 -> 246

ajs124 commented:
Reading through the changelog, this might require a change in some module: 
```
nss-mymachines lost support for resolution of users and groups, and
          now only does resolution of hostnames. This functionality is now
          provided by nss-systemd. Thus, the 'mymachines' entry should be
          removed from the 'passwd:' and 'group:' lines in /etc/nsswitch.conf
          (and 'systemd' added if it is not already there).
```

dasJ commented:
@ajs124 These are the corresponding lines: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/systemd.nix#L902,L916

flokli commented:
We also should review #91960 before merging this, rebase this PR on top and update the options in there.

flokli committed:
gnupgMinimal: move to all-packages.nix, use from inside systemd

This makes things more consistent with how we pass in utillinuxMinimal.

There's already a minimal argument in the gnupg derivation, but this
gnupg is even more minimal. Instead of defining a gnupgSuperMinimal, we
define it inline inside systemd.

flokli committed:
systemd: use tags as rev, update comment explaining updating workflow.

flokli commented:
I rebased this on latest staging, and updated the documentation regarding on how to do updates as suggested by @andersk in https://github.com/NixOS/nixpkgs/pull/94681#discussion_r465317254.

Still haven't yet looked into how to fix the cyclic dependency between the different outputs - help appreciated.

flokli committed:
nixos/systemd: remove mymachines nss module from passwd: and group: lines

From the systemd release notes:

nss-mymachines lost support for resolution of users and groups, and now
only does resolution of hostnames. This functionality is now provided by
nss-systemd. Thus, the 'mymachines' entry should be removed from the
'passwd:' and 'group:' lines in /etc/nsswitch.conf (and 'systemd' added
if it is not already there).

andir commented:
> $ grep -r /nix/store/sqsrif6i3ahmfypflwcn2mvd6n7yxbqw-systemd-246-lib /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246
> Binary file /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/libnss_systemd.so.2 matches
> Binary file /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/libsystemd-shared-246.so matches
> Binary file /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/security/pam_systemd.so matches

Those libs already existed in earlier versions but now (probably due to
some refactorings within systemd) have strings referring to the $lib
output:

> $ strings -f /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/libnss_systemd.so.2 /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/libsystemd-shared-246.so /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/security/pam_systemd.so | grep 246-lib
> /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/libnss_systemd.so.2: /nix/store/sqsrif6i3ahmfypflwcn2mvd6n7yxbqw-systemd-246-lib/lib/libnss_systemd.so.2
> /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/libsystemd-shared-246.so: /nix/store/sqsrif6i3ahmfypflwcn2mvd6n7yxbqw-systemd-246-lib/lib/libnss_systemd.so.2
> /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/libsystemd-shared-246.so: /nix/store/sqsrif6i3ahmfypflwcn2mvd6n7yxbqw-systemd-246-lib/lib
> /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/security/pam_systemd.so: /nix/store/sqsrif6i3ahmfypflwcn2mvd6n7yxbqw-systemd-246-lib/lib/libnss_systemd.so.2

- Why do we have libnss_systemd.so.2 both on $out/lib and in $lib/lib?
- We are referring to $lib in $out as you can see above.
- We are referring to $out in $lib in e.g. $lib/lib/libsystemd.so.0.29.0:
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/bin/
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/binfmt.d
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/modules-load.d
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/sysctl.d
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/network
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/system
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/system-generators
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/system-preset
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/system-shutdown
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/system-sleep
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/user
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/user-generators
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/systemd/user-preset
  /nix/store/5wymqk84lk71wbfhlrnk1z6y43cvrr5h-systemd-246/lib/sysusers.d

  These should probalby be moved into the lib output or point towards
  /run/current-system/systemd/lib/systemd/{network,system,system-generators,system-preset,…}
  instead. ? Maybe it is time to rethink the entire split output for
  systemd. It has been a mess for a while and currently everything is
  potentially referring to everything else.

flokli commented:
I did look a bit at this with @arianvp. `libsystemd-shared.so` contains some code used by many of the systemd binaries. It itself uses `libsystemd.so` (which is in the `lib` output). I think this is fine.

What's not fine (and seems to have changed in 9c5bb2033db0c0927593b929ff760b69295a9f70) is that `libsystemd.so` now contains a `get_path` helper function containing references back to ROOTPREFIX.

I'm not yet sure on whether these are used (and simply not garbage collected), similar to the [situation on aarch64](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/systemd/default.nix#L269).

flokli commented:
Apparently these paths are in the library are [now also exposed](https://github.com/systemd/systemd/commit/9c5bb2033db0c0927593b929ff760b69295a9f70) by [`sd_path_lookup*`](https://www.freedesktop.org/software/systemd/man/sd_path_lookup.html), which seems public libsystemd API :-/ (thanks @jtojnar)

I'm not really if it's wise to put all this logic into a generic library other applications link against, if it still contains hardcoded paths - IMHO, these applications should either ask the systemd `.pc` files during build-time, or if this can change at runtime, the libsystemd API should ask the running systemd for the paths currently active on the system, by asking pid 1 via IPC.

I'll take a look where this method is used inside of systemd itself, and whether we can/want to point these paths to something else (either by changing meson flags, or by patching the code). It seems some of the code added in above commit doesn't properly honor the path structure anyways, but uses some hardcoded strings in first place - maybe we should have a discussion with upstream about how to structure these paths.

If this all doesn't work out, we might have to give up the multi-output nature of systemd (and ship everything in a bigger single output).

arianvp commented:
Or we patch systemd and make these functions always return `/run/current-system/systemd`  instead of `ROOTPREFIX=` ?

flokli commented:
It depends how these are used inside systemd as well. systemd itself should probably still be able to find the correct locations to its binaries while switching from one generation to another, or while it mounts `/run/current-system` by itself (in case we do that with systemd-in-initrd)

flokli committed:
systemd: remove lib output

systemd 246 started baking in some strings into $lib/lib/libsystemd.so
which points back to $out.

It might be desirable to not have these strings in there, but instead
have systemd ask the running manager for the really active paths via
IPC, or patch in some more mutable paths (like /run/current-system/…) -
but as long as it is as it is, we should probably should just stop
producing the `lib` output for now.

This does increase the runtime closure for installations not pulling in
systemd.out, or linking against multiple libsystemd's, but doesn't
increase the runtime closure for NixOS itself, as it needs systemd
anyways, and doesn't link against multiple variants of it.

We can also remove the aarch64-specific $out/lib/systemd/catalog hack,
and revive it at a later time if the `lib` output is reintroduced.

flokli committed:
nixos/systemd: don't try to install systemd-initctl.{service,socket}

These are now only installed by systemd if HAVE_SYSV_COMPAT is true,
which only is the case if you set sysvinit-path and sysvrcnd-path (which
we explicitly unset in the systemd derivation for quite some time)

flokli committed:
nixos/boot: handle systemd-udevd being a symlink to udevadm

flokli commented:
I removed the `lib` output for now (with a detailled explanation in the commit message), and did some basic fixups and smoke testing - currently rebuilding my system for it. Please take a look :-)

flokli committed:
axoloti: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
brackets: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
lighttable: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
pencil: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
obinskit : use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
vscode: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
chromium: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
opera: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
drawio: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
ungoogled-chromium: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
signal-desktop: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
teams: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
thunderbird: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
mullvad-vpn: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
citrix-workspace: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
qt5.qtwebengine: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
python3.pkgs.pyudev: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
nwjs_0_12: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
nwjs: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
sysprof: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
filebeat6: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
filebeat7: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
loki: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
postfix-exporter: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
nixpart_0_4: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
staruml: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
datadog-agent: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
encryptr: use lib.getLib systemd instead of systemd.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
wire-desktop: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
simplenote: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
bluejeans: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
franz: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
rambox: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
vk-messenger: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
wavebox: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
mailspring: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
termius: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
cypress: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
hakuneko: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
bitwarden: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
keybase: use lib.getLib udev instead of udev.lib

This will pick the `lib` output if it exists, otherwise default to `out`.

flokli committed:
nixosTests.systemd: update output

systemd shows minutes as `min`, not `m`.

flokli commented on nixos/tests/systemd.nix:
```diff
@@ -145,8 +145,8 @@ import ./make-test-python.nix ({ pkgs, ... }: {
 
         output = machine.succeed("systemctl show | grep Watchdog")
         assert "RuntimeWatchdogUSec=30s" in output
```
For some reason this value isn't actually set, but stays at `0` (the others do get set).

I'm not sure if this is an upstream bug or something only happening inside the VM - maybe someone can take a closer look.

Emantor replied:
https://github.com/systemd/systemd/issues/15726 & https://github.com/systemd/systemd/pull/15932 may be relevant here.

flokli replied:
@Emantor this is only for per-service watchdog properties, which can be changed via the `sd_notify` protocol, isn't it?

Here, it's specifically the system-wide `RuntimeWatchdogUSec` property which doesn't seem to be honored, `RebootWatchdogUSec` and `KExecWatchdogUSec` are changed, which is really confusing.

vcunat commented:
> It depends how these are used inside systemd as well

The worst part is that we'd best re-check such things on every larger systemd update (if we do patch the path).

flokli commented:
Yeah, I planned to talk more with upstream and see if we can get some of the path handling there unified. For now it's probably fine to drop the `lib` output.

This PR probably just needs some more thorough testing in everyday usage, and someone toying with the watchdog options.

FRidh commented:
What's the status of this PR?

flokli commented:
I ran this on my laptop for a week or two and didn't spot any issues. I haven't figured yet out why the watchdog part of the test is failing.

I propose commenting this out and then merging this in.

I'd really like to see this in 20.09, and we'd still need to have it go through a staging round.

FRidh commented:
Right. I am about to start a new staging-next iteration which most likely would be the last one.

flokli commented:
Alright, let me fix the test and then merge this in.

flokli commented:
There will likely be some follow-up PRs to expose new options in the module system, but that's no world rebuild, so should be fine.

flokli committed:
nixosTests.systemd: disable RuntimeWatchdogUSec=30s assertion

For some reason, this value isn't updated, at least not inside the VM.

Uncomment it, so we still test the rest. Needs to be investigated
further.

andersk commented:
This broke fwupd: #96663.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/marketing-team-can-we-present-nix-nixos-better/6249/106
